app/Run.hs

```diff
@@ -545,53 +504,240 @@ mkChainDbArgs TraceOptions {traceChainDB} cfg initLedger registry (CoreNodeId ni
+    , tracerFetchDecisions
+        = enableTracer (traceFetchDecisions traceOptions)
+        $ withName "FetchDecision" tracer
+    , tracerFetchClient
+        = enableTracer (traceFetchClient traceOptions)
+        $ withName "FetchClient" tracer
+    , tracerTxInbound
+        = enableTracer (traceTxInbound traceOptions)
+        $ withName "TxInbound" tracer
+    , tracerTxOutbound
+        = enableTracer (traceTxOutbound traceOptions)
+        $ withName "TxOutbound" tracer
+    , traceIpSubscription
+        = withName "IPSubscription" tracer
+    , traceDnsSubscription
+        = withName "DNSSubscription" tracer
+    , traceDnsResolver
+        = withName "DNSResolver" tracer
+    }
+  where
+    tracer' :: Tracer IO String
+    tracer' = contramap pack $ toLogObject tracer
+
+    enableTracer
+      :: Show a
+      => Bool
+      -> Tracer IO String
+      -> Tracer IO a
+    enableTracer False = const nullTracer
+    enableTracer True  = showTracing
+
+    mempoolTraceTransformer :: Tracer IO (LogObject a)
+                            -> Tracer IO (TraceEventMempool blk)
+    mempoolTraceTransformer tr = Tracer $ \mempoolEvent -> do
+        let logValue = LogValue "txsInMempool" $ PureI $ fromIntegral $ _txsInMempool mempoolEvent
+        meta <- mkLOMeta Info Confidential
+        traceNamedObject tr (meta, logValue)
+        case mempoolEvent of
+          TraceMempoolAddTxs      txs _ ->
+              let logValue' = LogValue "txsProcessed" $ PureI $ fromIntegral $ length txs in
+              traceNamedObject tr (meta, logValue')
+          TraceMempoolRejectedTxs txs _ ->
+              let logValue' = LogValue "txsProcessed" $ PureI $ fromIntegral $ length txs in
+              traceNamedObject tr (meta, logValue')
+          _                             -> return ()
+
+
+
+--
+-- Tracing utils
+--
+
+withName :: String
+         -> Tracer IO (LogObject Text)
+         -> Tracer IO String
+withName name tr = contramap pack $ toLogObject $ appendName (pack name) tr
+
+-- | Tracing wrapper which includes current tip in the logs (thus it requires
+-- it from the context).
+--
+data WithTip blk a =
+    WithTip
+      (Point blk)
+      -- ^ current tip point
+      a
+      -- ^ data
+
+instance ( Show a
+         , Condense (HeaderHash blk)
+         ) => Show (WithTip blk a) where
+
+    show (WithTip tip a) = case pointHash tip of
+      GenesisHash -> "[genesis] " ++ show a
+      BlockHash h -> mconcat [ "["
+                             , take 7 (condense h)
+                             , "] "
+                             , show a
+                             ]
+
+
+-- | A way to satisfy tracer which requires current tip.  The tip is read from
+-- a mutable cell.
+--
+withTip :: TVar IO (Point blk)
+        -> Tracer IO (WithTip blk a)
+        -> Tracer IO a
+withTip varTip tr = Tracer $ \msg -> do
+    tip <- atomically $ readTVar varTip
+    traceWith (contramap (WithTip tip) tr) msg
```

Comment:
I think we need a TODO to do this differently. The consensus tracing can return this directly, and importantly can grab it in a way that's consistent with the event. Grabbing it in a separate STM transaction risks reporting different info.

Reply:
Making this work here was a bit tricky, and moving this to `ouroboros-consensus` would simplify things.  I'll add a TODO.

^ @mrBliss